server/src/main/java/org/elasticsearch/search/aggregations/metrics/HyperLogLogPlusPlus.java

```diff
@@ -179,17 +263,68 @@ void upgradeToHll(long bucket) {
         }
     }
 
+    // protected for testing
+    void upgradeToLinearCounting(long bucket, long ord) {
+        low.bucket = bucket;
+        lc.ord = ord;
+        hll.ensureCapacity(ord + 1);
+        final AbstractLinearCounting.HashesIterator hashes = low.values();
+        while (hashes.next()) {
+            lc.addEncoded(hashes.value());
+        }
+    }
+
+    // used for deserializing sketches
+    private void collectEncoded(long bucket, int encoded) {
+        if (algorithm.get(bucket) == LINEAR_COUNTING) {
+            low.bucket = bucket;
+            if (low.exhausted()) {
+                lc.ord = bucketToOrd(bucket);
+                if (lc.addEncoded(encoded) > lc.threshold) {
+                    upgradeToHll(bucket, lc.ord);
+                }
+            } else {
+                low.addEncoded(encoded);
+                if (low.exhausted()) {
+                    upgradeToLinearCounting(bucket, bucketToOrd(bucket));
+                }
+            }
+        } else {
+            hll.ord = bucketToOrd(bucket);
+            hll.collectEncoded(encoded);
+        }
+    }
+
+    private long bucketToOrd(long bucket) {
+        long size = ordMapping.size();
+        if (size <= bucket) {
+            ordMapping = bigArrays.grow(ordMapping, bucket + 1);
+            ordMapping.fill(size, ordMapping.size(), -1);
```

Comment:
Do we need to fill at all, I wonder? I *think* we only set this when we're "upgrading" into a proper linear counting so we can explicitly set it then. If that is the case we can skip the length check when we're just translating.